Add normative text to verification and processing section to verify public key (#175)
oauth-wg/oauth-sd-jwt-vc: draft-ietf-oauth-sd-jwt-vc.md

according to Section 6 of [@!I-D.ietf-oauth-selective-disclosure-jwt]. To verify
the Key Binding JWT, the `cnf` claim of the SD-JWT MUST be used.

> Furthermore, the recipient of the SD-JWT VC MUST verify the public key used to
> verify the Issuer-signed JWT corresponds to Issuer of the Issuer-signed JWT.

Reviewer: ```suggestion
Furthermore, the recipient of the SD-JWT VC MUST verify that the public key used for
verification of the Issuer-signed JWT belongs to the Issuer of the Issuer-signed JWT.
```

Reviewer: Please add a reference to {#verifying-public-key-for-issuer-signed-jwts}

Author: will add that reference

> Furthermore, the recipient of the SD-JWT VC MUST obtain the public verification key
> of the Issuer-signed JWT as defined (#public-key-discovery-for-issuer-signed-jwts).

Reviewer: ```suggestion
Furthermore, the recipient of the SD-JWT VC MUST obtain the public verification key
for the Issuer-signed JWT as defined in (#public-key-discovery-for-issuer-signed-jwts).
```

Additional validation rules MAY apply, but their use is out of the scope of this
specification.

## Obtaining Public Key for Issuer-signed JWTs {#public-key-discovery-for-issuer-signed-jwts}

> A recipient of an SD-JWT VC MUST apply the following rules to obtain the public
> verification key for the Issuer-signed JWT:

Reviewer: ```suggestion
A recipient of an SD-JWT VC MUST apply the following rules to obtain the public
verification key for the Issuer-signed JWT:

```

Additional validation rules MAY apply, but their use is out of the scope of this
specification.

## Verifying Public Key for Issuer-signed JWTs {#verifying-public-key-for-issuer-signed-jwts}

The following methods are RECOMMENDED to ensure the public key to verify the
Issuer-signed JWT corresponds to the Issuer:

- JWT Issuer Metadata: Obtain the public key using JWT Issuer Metadata as defined in (#jwt-issuer-metadata).
- DID Document Resolution: Retrieve the public key from the DID Document resolved from the DID in the `iss` value.
If the `kid` JWT header parameter is present, the `kid` MUST be a relative or absolute
DID URL of the DID in the `iss` value, identifying the public key.
- X.509 Certificates: Obtain the public key from the leaf X.509 certificate defined by the
`x5c`, `x5c`, or `x5t` JWT header parameters of the Issuer-signed JWT and validate the X.509
certificate chain. In this case, additionally the `iss` value of the Issuer-signed JWT MUST be
    - a DNS name that matches a `dNSName` Subject Alternative Name (SAN) [RFC5280] entry of

> the leaf certificate,
>     - or a URN that matches a `unifiedResourceName` SAN entry of the leaf certificate.

Reviewer: ```suggestion
the leaf certificate, or
    - a URN that matches a `unifiedResourceName` SAN entry of the leaf certificate.
```

Other methods for verifying the public key belongs to the Issuer MAY be used but are

> out-of-scope of this specification.

Reviewer: ```suggestion
out of scope of this specification.
```

Additional validation rules MAY apply, but their use is out of the scope of this
specification.

## Public Key for Issuer-signed JWTs Verification {#public-key-for-issuer-signed-jwts-verification}

> A recipient of the SD-JWT VC MUST apply the following rules to verify the
> Issuer-signed JWT corresponds to the Issuer:

Reviewer: ```suggestion
A recipient of the SD-JWT VC MUST use the first applying strategy of the following list to verify the
Issuer-signed JWT corresponds to the Issuer:
```

Author: Why is first applying needed. the options are mutually exclusive, aren't they?

- JWT Issuer Metadata: If the `iss` value contains an HTTPS URI, the recipient MUST
obtain the public key using JWT Issuer Metadata as defined in (#jwt-issuer-metadata).
- DID Document Resolution: If the `iss` value contains a DID, the recipient MUST retrieve
the public key from the DID Document resolved from the DID in the `iss` value.
If the `kid` JWT header parameter is present, the `kid` MUST be a relative or absolute
DID URL of the DID in the `iss` value, identifying the public key.

> - X.509 Certificates: The recipient MUST obtain the public key from the leaf X.509 certificate
> defined by the `x5c`, `x5c`, or `x5t` JWT header parameters of the Issuer-signed JWT and validate the X.509
> certificate chain in the following cases:

Author: ```suggestion
- X.509 Certificates: If the `x5c`, `x5c`, or `x5t` JWT header parameters are present, the recipient MUST obtain the public key from the leaf X.509 certificate defined by of the Issuer-signed JWT and validate the X.509 certificate chain in the following cases:
```

>     - If the `iss` value contains a DNS name encoded as a URI using the DNS URI scheme [@RFC4501]. The DNS name MUST matche a `dNSName` Subject Alternative Name (SAN) [@RFC5280] entry of the leaf certificate.

Reviewer: ```suggestion
    - If the `iss` value contains a DNS name encoded as a URI using the DNS URI scheme [@RFC4501]. In this case, the DNS name MUST match a `dNSName` Subject Alternative Name (SAN) [@RFC5280] entry of the leaf certificate.
```

>     - If the `iss` value contains a DNS name encoded as a URI using the DNS URI scheme [@RFC4501]. The DNS name MUST matche a `dNSName` Subject Alternative Name (SAN) [@RFC5280] entry of the leaf certificate.

Author: ```suggestion
    - If the `iss` value contains a DNS name encoded as a URI using the DNS URI scheme [@RFC4501]. The DNS name MUST match a `dNSName` Subject Alternative Name (SAN) [@RFC5280] entry of the leaf certificate.
```

>     - If the `iss` value contains a URN using the URN URI scheme [@RFC2141]. The URN MUST match a `unifiedResourceName` SAN entry of the leaf certificate.

Reviewer: ```suggestion
    - If the `iss` value contains a URN using the URN URI scheme [@RFC2141]. In this case, the URN MUST match a `unifiedResourceName` SAN entry of the leaf certificate.
```

Author: Accepted

> Additional rules for verifying the public key belongs to the Issuer MAY be used but are
> out of scope of this specification.

Reviewer: ```suggestion
Separate specifications or ecosystem regulations MAY define rules substituting or complementing the rules defined above, but such rules are
out of scope of this specification. See (#ecosystem-verification-rules) for security considerations.
```

Author: I would disagree with this. I don't think that ecosystems should introduce breaking changes. They MAY add rules on top of what was defined. Additionally, it is always a verifier decision whether or not a public key verification method is acceptable. This is a policy decision. Therefore I would keep the language. Does this make sense?

Author: The current text does not mean that verifiers have to blindly accept SD-JWT VCs with specific verification methods if it is against their policy.

Reviewer: I'm fine with removing the "substituting or" part. I don't think we should keep the "MAY be used" part in the original text as it opens to door quite a bit too wide.

Author: Sounds good. I'll adjust the text.

Additional considerations can be found in [@OWASP_SSRF].

> ## Ecosystem-specific Public Key Verification Methods

Reviewer: ```suggestion
## Ecosystem-specific Public Key Verification Methods {#ecosystem-verification-rules}
```

When defining ecosystem-specific rules for the verification of the public key,
as outlined in (#public-key-for-issuer-signed-jwts-verification), it is critical
that those rules maintain the integrity of the relationship between the `iss` value
within the Issuer-signed JWT and the public keys of the Issuer.

> In situations where an attacker tampers with an SD-JWT VC, it is paramount
> to ensure that they cannot exert influence over the verification process while
> employing the same `iss` value.

Reviewer: ```suggestion
It MUST be ensured that for any given `iss` value, an attacker cannot influence what
kind of verification process is used. Otherwise, an attacker could attempt to
have the verifier use a verification process that was not intended by the Issuer
and where the attacker may have the opportunity to influcence the verification result
to his advantage.
```

Author: How about...

```
It MUST be ensured that for any given `iss` value, an attacker cannot influence the type of verification process used. Otherwise, an attacker could attempt to make the Verifier use a verification process not intended by the Issuer, allowing the attacker to potentially manipulate the verification result to their advantage.
```

Author: I used my suggested language in the updated PR.
